libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/tasks/Tasks.kt

```diff
@@ -786,14 +797,29 @@ abstract class KotlinCompile @Inject constructor(
         return super.source(*sources)
     }
 
-    private fun getClasspathChanges(): ClasspathChanges {
-        val currentSnapshotFiles = classpathSnapshotProperties.classpathSnapshot.files.toList()
+    private fun getClasspathChanges(knownChangedFiles: ChangedFiles.Known): ClasspathChanges {
+        // Find current snapshot files that have been changed (added or modified)
+        val currentSnapshotFiles = classpathSnapshotProperties.classpathSnapshot.files
+        val addedOrModifiedFiles = knownChangedFiles.modified.toSet()
```

Comment:
Can we use Gradle APIs to the the list of changes directly?

Reply:
Done - I've used `InputChanges` now. But as we discussed offline, comparing the changed parts only will be incorrect if the classpath contains duplicate classes, so I'm comparing the entire classpath now to be safe. I'll try to address this in the next PR.